spock-core/src/main/java/org/spockframework/runtime/SimpleFeatureNode.java

```diff
@@ -29,62 +33,34 @@ public Type getType() {
 
   @Override
   public SpockExecutionContext prepare(SpockExecutionContext context) throws Exception {
-    return delegate.prepare(
-      context.withCurrentFeature(getNodeInfo())
-      //.withParentId(getUniqueId())
-    );
-  }
-
-  @Override
-  public SpockExecutionContext before(SpockExecutionContext context) throws Exception {
-    context = super.before(context);
-    ErrorInfoCollector errorInfoCollector = new ErrorInfoCollector();
-    context = context.withErrorInfoCollector(errorInfoCollector);
-    context.getRunner().runSetup(context);
-    errorInfoCollector.assertEmpty();
-
-    return context;
-  }
-
-  @Override
-  public void around(SpockExecutionContext context, Invocation<SpockExecutionContext> invocation) {
-    // Wrap the Feature invocation around the invocation of the Iteration delegate
-    super.around(context, ctx -> delegate.around(ctx, invocation));
+    context = super.prepare(context);
+    return context.withCurrentFeature(getNodeInfo());
   }
 
   @Override
   public SpockExecutionContext execute(SpockExecutionContext context, DynamicTestExecutor dynamicTestExecutor) throws Exception {
     verifyNotSkipped(getNodeInfo());
-    delegate.execute(context, dynamicTestExecutor);
-    return context;
-  }
 
-  @Override
-  public void after(SpockExecutionContext context) throws Exception {
-    ErrorInfoCollector errorInfoCollector = new ErrorInfoCollector();
-    context = context.withErrorInfoCollector(errorInfoCollector);
-    delegate.after(context);
-    // First the iteration node, then the Feature node
-    errorInfoCollector.assertEmpty();
-    super.after(context);
-  }
-
-  @Override
-  public void nodeFinished(SpockExecutionContext context, TestDescriptor testDescriptor, TestExecutionResult result) {
-    delegate.nodeFinished(context, testDescriptor, result);
-    super.nodeFinished(context, testDescriptor, result);
-  }
-
-  @Override
-  public void nodeSkipped(SpockExecutionContext context, TestDescriptor testDescriptor, SkipResult result) {
-    // Skipping this Feature implies that the Iteration is also skipped
-    delegate.nodeSkipped(context, testDescriptor, result);
-    super.nodeSkipped(context, testDescriptor, result);
-  }
-
-  @Override
-  public SkipResult shouldBeSkipped(SpockExecutionContext context) throws Exception {
-    return delegate.shouldBeSkipped(context);
+    AtomicReference<Throwable> result = new AtomicReference<>();
+    EngineExecutionListener executionListener = new EngineExecutionListener() {
+      @Override
+      public void executionSkipped(TestDescriptor testDescriptor, String reason) {
+        result.set(new TestAbortedException(reason));
+      }
+
+      @Override
+      public void executionFinished(TestDescriptor testDescriptor, TestExecutionResult testExecutionResult) {
+        testExecutionResult.getThrowable().ifPresent(result::set);
+      }
+    };
+
+    addChild(delegate);
+    dynamicTestExecutor.execute(delegate, executionListener);
```

Comment:
🤔 this can affect the threading behavior in parallel execution mode, it is quite possible that the delegate gets executed in a different thread.

Reply:
Of course, but what's the problem with that? It's basically the same as with a rolled-up data-driven feature.